api/src/pcapi/core/users/external/__init__.py

```diff
@@ -203,10 +207,41 @@ def get_user_attributes(user: User) -> UserAttributes:
     )
 
 
-def _get_bookings_categories_and_subcategories(user_bookings: list[Booking]) -> Tuple[list[str], list[str]]:
-    booking_subcategories_ids = list(set(booking.stock.offer.subcategoryId for booking in user_bookings))
-    booking_categories_ids = list(set(booking.stock.offer.subcategory.category_id for booking in user_bookings))
-    return booking_categories_ids, booking_subcategories_ids
+def _get_bookings_categories_and_subcategories(
+    user_bookings: list[Booking],
+) -> Tuple[list[str], list[str], Optional[str]]:
+    """
+    Returns a tuple of three values:
+    - the list of all distinct categories in user bookings (not canceled)
+    - the list of all distinct subcategories in user bookings (not canceled)
+    - the single most booked subcategory in user bookings:
+       * category in which the user has the highest number of bookings
+       * in case of several most booked subcategories, the one with the highest credit spent by the user
+    """
+    bookings_by_subcategories = defaultdict(list)
+    for booking in user_bookings:
+        if booking.status != BookingStatus.CANCELLED:
+            bookings_by_subcategories[booking.stock.offer.subcategoryId].append(booking)
+
+    sorted_subcategories_items = sorted(
+        bookings_by_subcategories.items(),
+        key=lambda key_value: (
+            len(key_value[1]),
+            sum([booking.stock.price for booking in key_value[1]]),
+        ),
+        reverse=True,
+    )
+
+    booking_subcategories_ids = [subcategory[0] for subcategory in sorted_subcategories_items]
+    booking_categories_ids = list(
+        set(
+            booking.stock.offer.subcategory.category_id
+            for booking in user_bookings
+            if booking.status != BookingStatus.CANCELLED
+        )
```

Comment:
on ne pourrait pas créer le set dans la première boucle for (lignes 222 à 224) ?